### Task 5: Sum square difference

As usual, we'll start with O(N) bruteforce:

In [8]:
import time 

start = time.time()

square_sum = 0
sum_of_squares = 0

for i in range(1, 101):
    square_sum += i
    sum_of_squares += i*i
    
square_sum = square_sum * square_sum

print(square_sum - sum_of_squares)

finish = time.time() - start
print("Time in seconds: ", finish)

25164150
Time in seconds:  0.00040030479431152344


But let's turn to a math background of a square of a sum:

pow(a+b) = pow(a) + pow(b) + 2ab

This works for any number of terms, so the difference we need:

* for N=2 is 2ab

* for N=4 is 2ab + 2ac + 2ad + 2bc + 2bd + 2cd

* for N=100 is 2(a(b+c+d+...) + b(c+d+)...)

Let's implement that:

In [28]:
def sm(n):
    return sum(range(1, n))


total = 0
N = 101

for i in range(1, N):
    total += (i * sm(i))

print(total * 2)

Which gives as the following results:   
25164150   
Time in seconds:  0.00034499168395996094   

But, unfortunately, this solution is even more complex, though it takes about O(n^2) speed, so let's try another way:

The formula of the natural numbers sum in [1,N] is (N*(N+1))/2    

And the formula of the sum of squares is (N(N+1)(2N+1))/6, so we just need to:

In [30]:
sm = 100 * (100+1)/ 2
squared = (100 * (100 + 1) * (2 * 100 + 1)) / 6

result = sm * sm - squared
 
print(result)

25164150.0


And this solution is also O(1), which is best-case perfomance for us.